dsp_interview_transcripts/pipeline/process_data.py

```diff
@@ -1,96 +1,301 @@
-    }).reset_index()
-    
+    grouped = (
+        df.groupby([conversation_col, "turn"])
+        .agg({"timestamp": "first", text_col: " ".join, role_col: "first", "uuid": "first"})
+        .reset_index()
+    )
+
     logger.info(f"Number of turns after concatenating consecutive roles: {len(grouped)}")
 
-    grouped = grouped.drop(columns=['turn'])
+    grouped = grouped.drop(columns=["turn"])
 
     return grouped
 
-def create_q_and_a_column(data_df, text_col = 'text_clean', conversation_col='conversation'):
-    """Concatenates each user message with the immediately preceding bot message.
-    In cases where the user has given a really short response, this should provide some
-    helpful context.
+
+def process_bot_qs(interviews_df: pd.DataFrame) -> Tuple[List[str], pd.DataFrame]:
+    """
+    Produces a list of unique sentences produced by the bot.
+    These will be matched to the interview guide.
+
+    Parameters:
+    - interviews_df (pd.DataFrame): DataFrame with bot conversation data.
+
+    Returns:
+    - Tuple[List[str], pd.DataFrame]: A list of unique sentences and DataFrame with exploded sentences.
     """
-    # Initialize a list to store the Q&A
-    q_and_a_list = []
-    
-    # Variable to keep track of the most recent BOT text
-    last_bot_text = ""
-    
-    df = data_df.copy()
-    
-    # Track the current conversation ID to know when it changes
-    current_conversation = None
-    
-    # Iterate over the rows of the DataFrame
-    for _, row in df.iterrows():
-        
-        # Check if the conversation has changed
-        if current_conversation != row[conversation_col]:
-            current_conversation = row[conversation_col]
-            last_bot_text = ""  # Reset the last bot text for a new conversation
-        
-        if row['role'] == 'BOT':
-            # Update the last bot text
-            last_bot_text = row[text_col]
-            q_and_a_list.append('')  # No Q&A for BOT rows
-        elif row['role'] == 'USER':
-            # Combine the last bot text and current user text if there is a preceding bot text,
-            # otherwise just use the user text
-            q_and_a_list.append(f"{last_bot_text}\n{row[text_col]}" if last_bot_text else row[text_col])
+    bot_qs = interviews_df[interviews_df["role"] == "BOT"]
+    # split into individual sentences so that if the original question is contained within the utterance,
+    # we have a better chance of catching it
+    bot_qs["sentences"] = bot_qs["text_clean"].apply(lambda x: sent_tokenize(x))
+    bot_qs = bot_qs.explode("sentences")
+    bot_qs_list = bot_qs["sentences"].unique().tolist()
+    return bot_qs_list, bot_qs
+
+
+def match_questions(bot_qs_list: List[str], questions: List[str], threshold: float = 0.85) -> pd.DataFrame:
+    """Match the input questions from our interview template,
+    and the actual questions produced by the bot, using cosine similarity.
```

Comment:
```suggestion
    and the actual sentences produced by the bot, using cosine similarity.
```

I guess you're not just matching questions but in fact all sentences?

Reply:
Hi, yes this is correct! Good spot